Examples/Algorithms/TrackFinding/src/TrackParamsEstimationAlgorithmFunction.cpp

```diff
@@ -0,0 +1,43 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/MagneticField/ConstantBField.hpp"
+#include "Acts/MagneticField/InterpolatedBFieldMap.hpp"
+#include "Acts/MagneticField/SharedBField.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+#include "ActsExamples/Plugins/BField/ScalableBField.hpp"
+#include "ActsExamples/TrackFinding/TrackParamsEstimationAlgorithm.hpp"
+
+namespace {
+
+template <typename bfield_t>
+struct BFieldGetterImpl {
+  bfield_t bField;
+
+  BFieldGetterImpl(bfield_t&& b) : bField(std::move(b)) {}
+
+  Acts::Vector3 operator()(const Acts::Vector3& position) const {
+    return bField.getField(position);
+  };
+};
+}  // namespace
+
+ActsExamples::TrackParamsEstimationAlgorithm::BFieldGetter
+ActsExamples::TrackParamsEstimationAlgorithm::makeBFieldGetter(
+    Options::BFieldVariant magneticField) {
+  return std::visit(
+      [](auto&& inputField) -> BFieldGetter {
+        using InputMagneticField =
+            typename std::decay_t<decltype(inputField)>::element_type;
+        using MagneticField = Acts::SharedBField<InputMagneticField>;
+        MagneticField field(std::move(inputField));
+
+        return BFieldGetterImpl<MagneticField>(std::move(field));
+      },
+      std::move(magneticField));
```

Comment:
Yeah, so it really seems like we need to think more about the B field infrastructure. This layers a ton of different type-erasure mechanisms on top of each other. The original reason why we don't use a single virtual interface for all B field providers is that we wanted to have the opaque cache object be a concrete type.

We **could** switch to using an `std::any` parameter in a generic interface, and then cast in the specific implementation after having gone through the vtable. This would eliminate the need for both the `std::variant` and the getter here, and in other spots.

Reply:
Sorry that I didn't think too much about this. It's really a lazy copy around. As @msmk0 suggested, the casting and the getter has been put into a simple private function now.

Reply:
This look good to me.